iris_grib/tests/unit/load_convert/test_product_definition_section.py

```diff
@@ -0,0 +1,134 @@
+# before importing anything else.
+import iris_grib.tests as tests
+
+import mock
+from iris.coords import DimCoord
+from iris_grib._load_convert import product_definition_section
+
+from iris_grib.tests.unit.load_convert import empty_metadata
+from iris_grib.tests.unit.load_convert.test_product_definition_template_0 \
+    import section_4 as template_0
+from iris_grib.tests.unit.load_convert.test_product_definition_template_31 \
+    import section_4 as template_31
+
+
+class Test(tests.IrisGribTest):
+    def setUp(self):
+        self.patch('warnings.warn')
+        self.translate_phenomenon_patch = self.patch(
+            'iris_grib._load_convert.translate_phenomenon'
+        )
+
+
+# Tests focussing on the handling of fixed surface elements in section 4.
+# Expects/ignores depending on the template number.
+class TestFixedSurfaces(Test):
+    def generic_fixed_surface_test(self, fs_is_expected, fs_is_present):
+        # Whether or not fixed surface elements are expected/present in the
+        # section 4 keys, most of the code is shared so we are using a single
+        # function with parameters.
+
+        # Prep placeholder variables for product_definition_section.
+        discipline = mock.sentinel.discipline
+        tablesVersion = mock.sentinel.tablesVersion
+        rt_coord = DimCoord(24, 'forecast_reference_time',
+                            units='hours since epoch')
+        metadata = empty_metadata()
+
+        # Use the section 4 from either product_definition_section #1 or #31.
+        # #1 contains fixed surface elements, #31 does not.
+        templates = {0: template_0(), 31: template_31()}
+        template_number = 0 if fs_is_expected else 31
+        section_4 = templates[template_number]
+        section_4.update({
+            'productDefinitionTemplateNumber': template_number,
+            'parameterCategory': None,
+            'parameterNumber': None
+        })
+
+        fixed_surface_keys = [
+            'typeOfFirstFixedSurface',
+            'scaledValueOfFirstFixedSurface',
+            'typeOfSecondFixedSurface'
+        ]
+        for key in fixed_surface_keys:
+            # Force the presence or absence of the fixed surface elements even
+            # when they're respectively ignored or expected.
+            if fs_is_present and key not in section_4:
+                section_4[key] = template_0()[key]
+            elif (not fs_is_present) and key in section_4:
+                del section_4[key]
+
+        def run_function():
+            # For re-use in every type of test below.
+            product_definition_section(
+                section_4, metadata, discipline, tablesVersion, rt_coord)
+
+        if fs_is_expected and not fs_is_present:
+            # Should error since the expected keys are missing.
+            with self.assertRaises(KeyError) as context:
+                run_function()
+            self.assertTrue('FixedSurface' in str(context.exception))
```

Comment:
@trexfeathers You should be able to replace lines 93-95 with:
```python
emsg = 'FixedSurface'
six.assertRaisesRegex(self, KeyError, emsg, run_function)
```
of course you'll also need to import `six`. Or if you prefer, use the context manager version e.g.
```python
with six.assertRaisesRegex(self, KeyError, emsg):
    run_function()
```

Reply:
It's purely looking for a key error where any of the fixed surface keys are missing, so the existing phrase works fine in this case. But I've switched the format as suggested, thanks for that